src/Arrays/Buffers.jl

```diff
@@ -0,0 +1,232 @@
+# with respect to other streams, but we need to synchronize the default
+# stream before the ISend
+###
+
+function prepare_transfer!(buf::Buffer)
+  if buf.transfer === nothing
+    # nothing to do here
+  else
+    copybuffer!(buf.transfer, buf.stage, async=false)
+  end
+end
+
+function prepare_stage!(buf::Buffer)
+  if buf.transfer === nothing
+    # nothing to do here
+  else
+    copybuffer!(buf.stage, buf.transfer, async=true)
+  end
+end
+
+######
+# Internal methods
+######
+register(x) = nothing
+unregister(x) = nothing
+
+"""
+  copybuffer(A, B; async=true)
+
+Copy a buffer from device to host or vice-versa. Internally this uses
+`cudaMemcpyAsync` on the `CuDefaultStream`. The keyword argument 
+`async` determines whether it is asynchronous with regard to the host.
+"""
+function copybuffer!(A::AbstractArray, B::AbstractArray; async=true)
+  copy!(A, B)
+end
+
+
+@init @require CUDAdrv = "c5f51814-7f29-56b8-a69c-e4d8f6be1fde" @require CuArrays = "3a865a2d-5b23-5a0f-bc46-62713ec82fae" begin
+  using .CUDAdrv
+  using .CuArrays
+
+  import CUDAdrv.Mem
+
+  function register(arr)
+    if sizeof(arr) == 0
+      @warn "Array is of size 0. Can't pin register array with CUDA" size(arr) typeof(arr)
+      return arr
+    end
+    GC.@preserve arr begin
+      # XXX: is Mem.HOSTREGISTER_DEVICEMAP necessary?
+      Mem.register(Mem.HostBuffer, pointer(arr), sizeof(arr), Mem.HOSTREGISTER_DEVICEMAP)
+    end
+  end
+
+  function unregister(arr)
+    if sizeof(arr) == 0
+      return
+    end
+    GC.@preserve arr begin
+      ptr = pointer(arr)
+      buf = Mem.HostBuffer(ptr, sizeof(arr), CuCurrentContext(), true)
+      # XXX: CUDAdrv.Mem.unregister is only implemented for HostBuffer
+      Mem.unregister(buf)
+    end
+  end
+
+
+  # CUDAdrv.jl throws on CUDA_ERROR_NOT_READY
+  function queryStream(hStream)
+      err = CUDAapi.@runtime_ccall((:cuStreamQuery, CUDAdrv.libcuda), CUDAdrv.CUresult,
+                          (CUDAdrv.CUstream,),
+                          hStream)
+
+      if err === CUDAdrv.CUDA_ERROR_NOT_READY
+          return false
+      elseif err === CUDAdrv.CUDA_SUCCESS
+          return true
+      else
+          CUDAdrv.throw_api_error(err)
+      end
+  end
+
+  """
+      friendlysynchronize(stream)
+
+  MPI defines a notion of progress which means that MPI operations
+  need the program to call MPI functions (potentially multiple times)
+  to make progress and eventually complete. In some implementations,
+  progress on one rank may need MPI to be called on another rank.
+
+  As a result blocking by for example calling cudaStreamSynchronize,
+  may create a deadlock in some cases because not calling MPI will
+  not make other ranks progress.
+  """
+  function friendlysynchronize(stream)
+    status = false
+    while !status
+      status = queryStream(stream)
+      MPI.Iprobe(MPI.MPI_ANY_SOURCE, MPI.MPI_ANY_TAG, MPI.COMM_WORLD)
+    end
+    return
+  end
+
```

Comment:
Interesting.  I have not run into this issue before.  Is it specific to CUDA aware MPI?  Or is it because you are trying to do multiple buffers with your BufferKind type?

For the codes we have written in the past we have used a pattern like the following

https://github.com/lcw/Heptapus.jl/blob/0f235861c201cbfdcfc4368629e1dccda3868b57/examples/mpicuda/try.jl#L66-L83

Reply:
> Is it specific to CUDA aware MPI?

Not really, generally any spinlock will block MPI from making progress. I am probably overly cautious here, but I was trying to figure out why bors was freezing.